### ISSUE TITLE
Support image-set on background-images

### ISSUE THREAD

**Issue by verlok** [verlok/vanilla-lazyload#543]
[`image-set` is coming](https://css-tricks.com/using-performant-next-gen-images-in-css-with-image-set/)

That will change the need of having two separate `data-bg` and `data-bg-hidpi` attributes.
A single data-bg attribute will do it all.

```js
<div class='lazy' data-bg='image-set("platypus.png" 1x, "platypus-2x.png" 2x);'>
...
</div>
```

That will do the trick, as soon as `image-set` will be supported by all browsers.

[Check browser support](https://caniuse.com/css-image-set).

As of Nov 2021 it's still not possible to replace `data-bg` and `data-bg-hidpi` because of limited browser support.

**Comment by silveoj** [comment]
So don't you consider to parse like this?

input:
```data-bg-set='"small.jpg" 1x, "medium.jpg" 2x, "large.jpg" 3x'```
or
```data-bg='image-set("platypus.png" 1x, "platypus-2x.png" 2x);'```

output:
```
style="
  background-image: url("small.jpg");
  background-image: -webkit-image-set(
    url("small.jpg") 1x,
    url("medium.jpg") 2x
    url("large.jpg") 3x
  );
  background-image: image-set(
    url("small.jpg") 1x,
    url("medium.jpg") 2x
    url("large.jpg") 3x
  );
"
```

**Comment by verlok** [comment]
Hey @silveoj 
good question!

I’ll have to take a further look on this. 
Will do it tomorrow. 

Thank you for bringing this up!

**Comment by verlok** [comment]
Actually, it currently looks like `image-set` is supported well-enough by all browsers!
So it could probably replace the need of using `data-bg` + `data-bg-hidpi`, and use only `data-bg` instead.

**Comment by chocolata** [comment]
Hi @verlok ,
Thanks for your excellent work. Could you let us know if you have a timeline for this features in mind? It would be great to be able to lazyload a portrait image on mobile and a landscape image on desktop. Especially for those large hero backgrounds. Thanks for looking into it.

**Comment by verlok** [comment]
By the way, @chocolata, do you think [the current browser support](https://caniuse.com/css-image-set) would be enough for your use case?

**Comment by verlok** [comment]
> It would be great to be able to lazyload a portrait image on mobile and a landscape image on desktop. Especially for those large hero backgrounds.

@chocolata how you would use `image-set` to do this?

**Comment by chocolata** [comment]
Most definitely. I think it would be unreasonable of us to expect lazyload to support something that isn't supported by all major browsers. If it could fall back to the first image if it's an unsupported browser, that would be perfect. Thank you!

**Comment by verlok** [comment]
[This](https://github.com/verlok/vanilla-lazyload/issues/543#issuecomment-1011239922) is a very good idea @silveoj, I think I will go with the first way

```html
data-bg-set='"small.jpg" 1x, "medium.jpg" 2x, "large.jpg" 3x'
```

which would become, once entered:

```html
style="
  background-image: url("small.jpg");
  background-image: -webkit-image-set(
    url("small.jpg") 1x,
    url("medium.jpg") 2x
    url("large.jpg") 3x
  );
  background-image: image-set(
    url("small.jpg") 1x,
    url("medium.jpg") 2x
    url("large.jpg") 3x
  );
"
```

A question for you though: is the `-webkit-` prefix required? And if yes, on which browsers?

**Comment by verlok** [comment]
@chocolata have you read [this message](https://github.com/verlok/vanilla-lazyload/issues/543#issuecomment-1108529516)?

**Comment by verlok** [comment]
Guys, I started the implementation on [this branch](https://github.com/verlok/vanilla-lazyload/tree/feature/image-set-support) and it works great, it even supports multiple background so one could use only `data-bg-set` and get rid of `data-bg`, `data-bg-hidpi`, `data-bg-multi` and `data-bd-multi-hidpi` in one shot.

BUT, it works only on Safari and Firefox, while it fails to load any image on Chromium browsers (tried Chrome and Edge).

I debugged and, while the following lines execute correctly:

```js
element.style.webkitBackgroundImage = `image-set(${value})`;
element.style.backgroundImage = `image-set(${value})`;
```

after the execution of these lines, no `style` is applied to the DOM elements.

Works like a charm on Firefox and Safari though.

Any hint of what it could be?
Should I file a bug to Chromium?

**Comment by verlok** [comment]
In the screenshot, from the left:

- Chrome
- Firefox
- Safari

<img width="1920" alt="image" src="https://user-images.githubusercontent.com/1127721/165106525-867f44dd-eaf7-44fd-92f6-6f1223a58ac5.png">

**Comment by verlok** [comment]
Fixed! 

This is the fix: setting `element.style.backgroundImage` and checking if its value is empty right after. If it is, set it again with the `-webkit-` prefix.

```js
    let bgImageValues = imgSetValues.map((value) => `image-set(${value})`);
    element.style.backgroundImage = bgImageValues.join();
    // Temporary fix for Chromium with the -webkit- prefix
    if (element.style.backgroundImage === '') {
        bgImageValues = imgSetValues.map((value) => `-webkit-image-set(${value})`);
        element.style.backgroundImage = bgImageValues.join();
    }
```

It works in every browser now!

<img width="1920" alt="image" src="https://user-images.githubusercontent.com/1127721/165113948-49ba50cf-8e53-4c92-a71e-f2417fe8f32c.png">

**Comment by verlok** [comment]
So guys, the new feature is now on its own branch, I just need to merge it to `develop`, write some documentation, then release it.

**Comment by chocolata** [comment]
> @chocolata have you read [this message](https://github.com/verlok/vanilla-lazyload/issues/543#issuecomment-1108529516)?

Hi, sorry, I was away from my desk. It seems I'm completely mistaken. From what I can see, image-set() doesn't seem to support different images based solely on viewport-width... 

Damn. It would have been great to provide different images based on different breakpoints using a similar way as 

```
.slide {
  background-image: url('portrait.jpg');
}
@media (min-width: 768px) { 
  .slide {
    background-image: url('landscape.jpg'); 
  }
}
```
But I only now see a way to do it based on resolution... :-( 
Does anyone have an idea how to tackle this?

Botched JsFiddle here:
[https://jsfiddle.net/nrvoLuz6/2/](https://jsfiddle.net/nrvoLuz6/2/)

**Comment by verlok** [comment]
@chocolata I think it would be a good idea to open a new issue, a new thread, asking for help about how to do that. Let’s leave this thread for `image-set` support then.

**Comment by verlok** [comment]
Just [released 17.8.1](https://github.com/verlok/vanilla-lazyload/releases/tag/17.8.1) with `image-set` support via `data-bg-set`.

Thank you @Spone @silveoj @chocolata and @G100g for your contributions!

If you’re happy with my support, the features, the documentation and generally the effort I’ve been putting on this project in the latest years, I hope you might want to [buy me a coffee](https://ko-fi.com/verlok) to show your appreciation. Or [sponsor me](https://github.com/sponsors/verlok), if you're a fan.

Open-source software is great for everyone, but it takes passion and time (and coffee!) to grow and evolve.

Thank you for thinking about it. 
Have a wonderful day!
